# 🏥 Healthcare Patient Analytics & Risk Analysis System

## Phase 2 — Day 4: Duplicate Record Handling

### Objective

The objective of Day 4 is to identify, inspect, and handle
duplicate patient records.

We will:

- Identify duplicate rows
- Count duplicate records
- Inspect duplicate records
- Remove exact duplicate rows
- Verify the cleaned dataset
- Save the processed dataset

In [4]:
# import libraries
import pandas as pd
import numpy as np

print("Libraries imported successfully.")

Libraries imported successfully.


In [6]:
# load dataset
df = pd.read_csv("/healthcare_patients.csv")

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [7]:
df.shape

(10020, 16)

In [10]:
# Check for duplicate rows
is_duplicated = df.duplicated()
is_duplicated.sum()

np.int64(20)

In [17]:
# Inspect the duplicate rows
duplicates = df[df.duplicated()]

duplicates


,patient_id,age,gender,blood_pressure,heart_rate,bmi,smoking_status,diabetes,hypertension,disease,admission_type,length_of_stay,medication_count,treatment_cost,readmission,risk_score
1553,102968,84,Male,127.0,84.0,24.1,Never,Yes,Yes,Respiratory Disease,Urgent,9,3.0,41759.0,No,74.8
2748,102247,42,Male,158.0,77.0,33.8,Never,No,No,Cancer,Urgent,3,1.0,26350.0,Yes,50.2
2784,103535,70,Female,142.0,86.0,28.1,Never,No,Yes,Hypertension,Elective,5,6.0,30611.0,No,37.1
3125,102552,33,Female,116.0,91.0,30.6,Never,No,No,Diabetes,Elective,3,2.0,16977.0,No,20.3
3160,101978,85,Male,125.0,83.0,23.5,Current,No,No,Healthy/Preventive,Elective,4,4.0,14079.0,No,51.5
4301,100404,24,Female,118.0,60.0,26.0,Never,No,No,Respiratory Disease,Elective,3,2.0,19725.0,No,12.1
5879,100053,71,Male,126.0,95.0,21.8,Never,Yes,No,Healthy/Preventive,Emergency,3,2.0,13236.0,No,59.0
5940,103881,49,Other,141.0,89.0,32.1,Current,No,Yes,Hypertension,Emergency,7,4.0,27534.0,No,62.0
6237,102442,43,Male,111.0,93.0,34.8,Current,No,No,Healthy/Preventive,Elective,1,4.0,6304.0,No,44.5
6542,103226,78,Female,145.0,67.0,30.2,Never,Yes,Yes,Heart Disease,Elective,10,3.0,47584.0,Yes,89.0


In [18]:
duplicates.shape

(20, 16)

In [25]:
# Check duplicate patient IDs
duplicated_id = df[df.duplicated()]['patient_id'].sum()

duplicated_id
df["patient_id"].duplicated().sum()

np.int64(20)

In [26]:
duplicate_patient_ids = df[
    df["patient_id"].duplicated(keep=False)
].sort_values("patient_id")

duplicate_patient_ids

,patient_id,age,gender,blood_pressure,heart_rate,bmi,smoking_status,diabetes,hypertension,disease,admission_type,length_of_stay,medication_count,treatment_cost,readmission,risk_score
3730,100053,71,Male,126.0,95.0,21.8,Never,Yes,No,Healthy/Preventive,Emergency,3,2.0,13236.0,No,59.0
5879,100053,71,Male,126.0,95.0,21.8,Never,Yes,No,Healthy/Preventive,Emergency,3,2.0,13236.0,No,59.0
6531,100271,84,Female,131.0,73.0,30.7,Never,No,No,Diabetes,Elective,3,0.0,9490.0,No,39.8
7466,100271,84,Female,131.0,73.0,30.7,Never,No,No,Diabetes,Elective,3,0.0,9490.0,No,39.8
4301,100404,24,Female,118.0,60.0,26.0,Never,No,No,Respiratory Disease,Elective,3,2.0,19725.0,No,12.1
3372,100404,24,Female,118.0,60.0,26.0,Never,No,No,Respiratory Disease,Elective,3,2.0,19725.0,No,12.1
3277,100602,47,Male,123.0,87.0,30.9,Never,No,No,Healthy/Preventive,Emergency,3,4.0,15928.0,No,36.2
9224,100602,47,Male,123.0,87.0,30.9,Never,No,No,Healthy/Preventive,Emergency,3,4.0,15928.0,No,36.2
8218,101699,39,Female,134.0,80.0,34.5,Former,No,Yes,Respiratory Disease,Elective,6,4.0,28421.0,No,44.1
5723,101699,39,Female,134.0,80.0,34.5,Former,No,Yes,Respiratory Disease,Elective,6,4.0,28421.0,No,44.1


In [27]:
# Confirm the duplicates are exact
df[df.duplicated(keep=False)].sort_values("patient_id")

,patient_id,age,gender,blood_pressure,heart_rate,bmi,smoking_status,diabetes,hypertension,disease,admission_type,length_of_stay,medication_count,treatment_cost,readmission,risk_score
3730,100053,71,Male,126.0,95.0,21.8,Never,Yes,No,Healthy/Preventive,Emergency,3,2.0,13236.0,No,59.0
5879,100053,71,Male,126.0,95.0,21.8,Never,Yes,No,Healthy/Preventive,Emergency,3,2.0,13236.0,No,59.0
6531,100271,84,Female,131.0,73.0,30.7,Never,No,No,Diabetes,Elective,3,0.0,9490.0,No,39.8
7466,100271,84,Female,131.0,73.0,30.7,Never,No,No,Diabetes,Elective,3,0.0,9490.0,No,39.8
4301,100404,24,Female,118.0,60.0,26.0,Never,No,No,Respiratory Disease,Elective,3,2.0,19725.0,No,12.1
3372,100404,24,Female,118.0,60.0,26.0,Never,No,No,Respiratory Disease,Elective,3,2.0,19725.0,No,12.1
3277,100602,47,Male,123.0,87.0,30.9,Never,No,No,Healthy/Preventive,Emergency,3,4.0,15928.0,No,36.2
9224,100602,47,Male,123.0,87.0,30.9,Never,No,No,Healthy/Preventive,Emergency,3,4.0,15928.0,No,36.2
8218,101699,39,Female,134.0,80.0,34.5,Former,No,Yes,Respiratory Disease,Elective,6,4.0,28421.0,No,44.1
5723,101699,39,Female,134.0,80.0,34.5,Former,No,Yes,Respiratory Disease,Elective,6,4.0,28421.0,No,44.1


In [31]:
# remove duplicates
df_cleaned = df.copy()

df_cleaned = df_cleaned.drop_duplicates()

df_cleaned.shape

(10000, 16)

In [33]:
# verify duplicate count
df_cleaned.duplicated().sum()

df_cleaned["patient_id"].duplicated().sum()

np.int64(0)

In [36]:
# Verify the original data wasn't changed
print("Original dataset:", df.shape)
print("Cleaned dataset:", df_cleaned.shape)

Original dataset: (10020, 16)
Cleaned dataset: (10000, 16)


In [38]:
# Save the processed dataset
import os

os.makedirs("/content/data/processed", exist_ok=True)


output_path = "/content/data/processed/healthcare_patients_no_duplicates.csv"

df_cleaned.to_csv(output_path, index=False)

print("Cleaned dataset saved successfully.")

Cleaned dataset saved successfully.


In [40]:
# verify the saved file
pd.read_csv(
    "/content/data/processed/healthcare_patients_no_duplicates.csv"
).shape

pd.read_csv(
    "/content/data/processed/healthcare_patients_no_duplicates.csv"
).duplicated().sum()

np.int64(0)

# 📋 Day 4 Observations

## Duplicate Records

The original dataset contained 10,020 records and 16 columns.

A total of 20 exact duplicate rows were identified.

## Duplicate Verification

The duplicate records were inspected before removal.

The repeated records were exact duplicates across the dataset
columns.

The duplicate patient IDs corresponded to these exact duplicate
records.

## Duplicate Handling

The 20 exact duplicate rows were removed using:

`drop_duplicates()`

The dataset size changed from:

10,020 rows → 10,000 rows

The number of columns remained unchanged at 16.

## Verification

After duplicate removal:

- Duplicate rows: 0
- Duplicate patient IDs: 0
- Final rows: 10,000
- Final columns: 16

## Data Preservation

The original raw dataset was not modified.

A separate processed dataset was created for further analysis.

## Conclusion

The exact duplicate records have been successfully identified
and removed.

The resulting dataset contains 10,000 unique patient records
and is ready for the next stage of data cleaning.

In [44]:
df.duplicated().sum()
df_cleaned.shape
df_cleaned.duplicated().sum()
df_cleaned["patient_id"].duplicated().sum()

np.int64(0)